src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+                        totalMoved += inserted;
+                        transferTransaction.commit();
+                    }
+                }
+
+                if (amount == totalMoved) {
+                    // early return if nothing can be moved anymore
+                    subTransaction.commit();
+                    return totalMoved;
+                }
+            }
+
+            subTransaction.commit();
+            return totalMoved;
+        } catch (Exception e) {
+            CrashReport report = CrashReport.forThrowable(e, "Moving resources between resource handlers");
+            //noinspection DataFlowIssue
+            report.addCategory("Move details")
+                    .setDetail("Input", from::toString)
+                    .setDetail("Output", to::toString)
+                    .setDetail("Filter", filter::toString)
+                    .setDetail("Amount", amount)
+                    .setDetail("Transaction", transaction);
+            throw new ReportedException(report);
+        }
+    }
+
+    /**
+     * Similar to {@link #move}, but instead of all resources, it will be the first one that matches the filter. While this won't be a full list, this will simplify things like {@link FluidUtil#moveFluidWithSound}
+     */
+    public static <R extends IResource, S> S moveFirstOrDefault(
+            @Nullable IResourceHandler<R> from,
+            @Nullable IResourceHandler<R> to,
+            Predicate<R> filter,
+            int amount,
+            R defaultResource,
+            @Nullable TransactionContext transaction,
+            IStackFactory<R, S> stackFactory) {
+        Objects.requireNonNull(filter, "Filter may not be null");
+
+        if (from == null || to == null)
+            return stackFactory.create(defaultResource, 0);
+
+        try {
+            int totalMoved = 0;
+            R lastMovedResource = defaultResource;
+
+            int size = from.size();
+
+            for (int index = 0; index < size; ++index) {
+                R fromResource = from.getResource(index);
+                if (fromResource.isEmpty() || !filter.test(fromResource)) continue;
+
+                // check how much can be extracted
+                int extracted;
+                try (Transaction simulatedExtractTransaction = TransactionManager.open(transaction)) {
+                    extracted = from.extract(index, fromResource, amount - totalMoved, simulatedExtractTransaction);
+                }
+
+                try (Transaction transferTransaction = TransactionManager.open(transaction)) {
+                    // check how much can be inserted
+                    int inserted = to.insert(fromResource, extracted, transferTransaction);
+
+                    // extract it, or rollback if the amounts don't match
+                    if (from.extract(index, fromResource, inserted, transferTransaction) == inserted) {
```

Comment:
Do we want to validate that `inserted > 0` here before bothering to extract, or should we just specify in the docs encouraging people to short circuit their impls if a zero amount is being inserted/extracted?